apps/taxes/src/pages/api/webhooks/order-calculate-taxes.ts

```diff
@@ -25,48 +40,34 @@ export default orderCalculateTaxesSyncWebhook.createHandler(async (req, res, ctx
   const { payload } = ctx;
   logger.info({ payload }, "Handler called with payload");
 
-  const validation = calculateTaxesPayloadSchema.safeParse(payload);
-
-  if (!validation.success) {
-    logger.error({ error: validation.error.format() }, "Payload is invalid");
+  try {
+    verifyCalculateTaxesPayload(payload);
+    logger.info("Payload validated succesfully");
+  } catch (error) {
+    logger.error({ error: error }, "Payload is invalid");
     logger.info("Returning no data");
-    return res.status(200).json({});
+    return res.send({});
   }
 
-  const { data } = validation;
-  logger.info("Payload validated succesfully");
-
-  const { providers, channels } = getAppConfig(data);
-  logger.debug("Parsed providers & channels from payload");
-
   try {
-    const channelSlug = data.taxBase.channel.slug;
-    const channelConfig = channels[channelSlug];
+    const appMetadata = payload.recipient?.privateMetadata ?? [];
+    const channelSlug = payload.taxBase.channel.slug;
+    const activeTaxProvider = getActiveTaxProvider(channelSlug, appMetadata);
 
-    if (!channelConfig) {
-      logger.error(`Channel config not found for channel ${channelSlug}`);
+    if (!activeTaxProvider.ok) {
+      logger.error({ error: activeTaxProvider.error }, "getActiveTaxProvider error");
       logger.info("Returning no data");
-      return res.send({});
+      return res.status(200).json({ success: false });
     }
 
-    const providerInstance = providers.find(
-      (instance) => instance.id === channelConfig.providerInstanceId
-    );
-
-    if (!providerInstance) {
-      logger.error(`Channel (${channelSlug}) providerInstanceId does not match any providers`);
-      logger.info("Returning no data");
-      return res.send({});
-    }
-
-    const taxProvider = new ActiveTaxProvider(providerInstance);
-    const calculatedTaxes = await taxProvider.calculate(data.taxBase, channelConfig);
+    logger.info({ activeTaxProvider }, "Fetched activeTaxProvider");
+    const taxProvider = activeTaxProvider.data;
+    const calculatedTaxes = await taxProvider.calculateTaxes(payload.taxBase);
 
     logger.info({ calculatedTaxes }, "Taxes calculated");
     return res.send(ctx.buildResponse(calculatedTaxes));
   } catch (error) {
     logger.error({ error }, "Error while calculating taxes");
-    logger.info("Returning no data");
-    return res.send({});
+    return res.status(400).json({ success: false, message: "Error while calculating taxes" });
```

Comment:
are you sure its always 400?

You should probably first validate if payload is correct. If not, return 400, then continue with calculation. If it fails, likely its not 400 anymore

Reply:
> You should probably first validate if payload is correct.

I am doing that in the first try/catch block.

>If not, return 400, then continue with calculation.

If I return an error because of an incorrect payload, I will retrigger the webhook. This is unwanted behavior because `X_CALCULATE_TAXES` webhooks get triggered even though the payload is insufficient to do tax calculations (f.e. when there are no line items or no address).

So when the payload is incorrect, I return the "success" response with no data.

In this try/catch, if something fails, it must be external. I agree there are better status codes than 400. What do you propose? 422? 500?

Reply:
I think it can be 500, message should be enough to debug here

For this `200` status, I understand, but this is "strange" behavior so please add comment to describe it